doc/specs/stdlib_hashmaps.md

```diff
@@ -0,0 +1,2088 @@
+store additional data associated with a key. Both types are
+opaque. Their current representations are as follows
+
+```fortran
+    type :: key_type
+        private
+        integer(int8), allocatable :: value(:)
+    end type key_type
+
+    type :: other_type
+        private
+        class(*), allocatable :: value
+    end type other_type
+```
+
+The  module also defines six procedures for those types: `copy_key`,
+`copy_other`, `equal_keys`, `free_key`, `free_other`, `get`, and
+`set`, and one operator, `==`,
+for use by the hash maps to manipulate or inquire of components of
+those types.
+
+### Table of `stdlib_hashmap_wrappers` procedures
+
+The  `stdlib_hashmap_wrappers` module provides procedures in
+several categories: procedures to manipulate data of the `key_type`;
+procedures to manipulate data of the `other_type`, and 32 bit hash
+functions for keys. The procedures in each category are listed
+below. It also provides an operator to compare two key type values for
+equality. 
+
+Procedures to manipulate `key_type` data:
+
+* `copy_key( key_in, key_out )` - Copies the contents of the key,
+  `key_in`, to contents of the key, `key_out`.
+
+* `get( key, value )` - extracts the contents of `key` into `value`,
+  an `int8` array or character string.
+
+* `free_key( key )` - frees the memory in `key`.
+
+* `set( key, value )` - sets the content of `key` to `value`.
+
+Procedures to manipulate `other_type` data:
+
+* `copy_other( other_in, other_out )` - Copies the contents of the
+  other data, `other_in`, to the contents of the other data,
+  `other_out`.
+
+* `get( other, value )` - extracts the contents of `other` into the
+  class(*) variable `value`.
+
+* `set( other, value )` - sets the content of `other` to the class(*)
+  variable `value`. 
+
+* `free_other( other )` - frees the memory in `other`.
+
+Procedures to hash keys to 32 bit integers:
+
+* `fnv_1_hasher( key )` - hashes a `key` using the FNV-1 algorithm.
+
+* `fnv_1a_hasher( key )` - hashes a `key` using the FNV-1a algorithm.
+
+* `seeded_nmhash32_hasher( key )` - hashes a `key` using the nmhash32
+  algorithm.
+
+* `seeded_nmhash32x_hasher( key )` - hashes a `key` using the nmhash32x
+  algorithm.
+
+* `seeded_water_hasher( key )` - hashes a `key` using the waterhash
+  algorithm.
+
+Operator to compare two `key_type` values for equality
+
+* `key1 == key2` - compares `key1' with 'key2' for equality
+
+### Specifications of the `stdlib_hashmap_wrappers` procedures
+
+#### `copy_key` - Returns a copy of the key
+
+##### Status
+
+Experimental
+
+##### Description
+
+Returns a copy of an input of type `key_type`.
+
+##### Syntax
+
+`call [[stdlib_hashmap_wrappers:copy_key]]( key_in, key_out )`
+
+##### Class
+
+Subroutine.
+
+##### Arguments
+
+`key_in`: shall be a scalar expression of type `key_type`. It
+is an `intent(in)` argument.
+
+`key_out`: shall be a scalar variable of type `key_type`. It
+is an `intent(out)` argument.
```

Comment:
What about `old_key` and `new_key` instead of `key_in` and `key_out`? I think it's important to pick variables names which "read" well, i.e "make a call to `copy_key`, passing an `old_key` as input, and receiving a `new_key` as output",

Reply:
I have done this, but wonder if `key_copy` Might be better.